server/adaptors/integrations/repository/integration_reader.ts

```diff
@@ -0,0 +1,619 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import path from 'path';
+import { validateTemplate } from '../validators';
+import { FileSystemCatalogDataAdaptor } from './fs_data_adaptor';
+import { CatalogDataAdaptor, IntegrationPart } from './catalog_data_adaptor';
+
+/**
+ * Helper method: Convert an Array<Result<type>> to Result<Array<type>>.
+ *
+ * @param results The list of results to fold.
+ * @returns A single result object with values in an array, or an error result.
+ */
+export const foldResults = <T>(results: Array<Result<T>>) =>
+  results.reduce(
+    (result, currentValue) => {
+      if (!result.ok) {
+        return result;
+      }
+      if (!currentValue.ok) {
+        return currentValue;
+      }
+      result.value.push(currentValue.value);
+      return result;
+    },
+    { ok: true, value: [] } as Result<T[]>
+  );
+
+/**
+ * Remove all fields from SerializedIntegration not present in IntegrationConfig.
+ *
+ * @param rawConfig The raw config to prune
+ * @returns A config with all data fields removed
+ */
+const pruneConfig = (rawConfig: IntegrationConfig | SerializedIntegration): IntegrationConfig => {
+  // Hacky workaround: we currently only need to prune 'data' fields, so just remove every 'data'.
+  // Lots of risky conversion in this method, so scope it to here and rewrite if more granular
+  // pruning is needed.
+  const prunePart = <T>(part: T): T => {
+    const result = {} as { [key: string]: unknown };
+    for (const [key, value] of Object.entries(part as { [key: string]: unknown })) {
+      if (key === 'data') {
+        continue;
+      } else if (Array.isArray(value)) {
+        result[key] = value.map((item) => {
+          if (item instanceof Object && item !== null) {
+            return prunePart(item);
+          }
+          return item;
+        });
+      } else if (value instanceof Object && value !== null) {
+        result[key] = prunePart(value as { [key: string]: unknown });
+      } else {
+        result[key] = value;
+      }
+    }
+    return (result as unknown) as T;
+  };
+
+  return prunePart(rawConfig);
+};
+
+/**
+ * Helper function to compare version numbers.
+ * Assumes that the version numbers are valid, produces undefined behavior otherwise.
+ *
+ * @param a Left-hand number
+ * @param b Right-hand number
+ * @returns -1 if a < b, 1 if a > b, 0 otherwise.
+ */
+export function compareVersions(a: string, b: string): number {
+  const aParts = a.split('.').map((part) => Number.parseInt(part, 10));
+  const bParts = b.split('.').map((part) => Number.parseInt(part, 10));
+
+  for (let i = 0; i < Math.max(aParts.length, bParts.length); i++) {
+    const aValue = i < aParts.length ? aParts[i] : 0;
+    const bValue = i < bParts.length ? bParts[i] : 0;
+
+    if (aValue > bValue) {
+      return 1;
+    }
+    if (aValue < bValue) {
+      return -1;
+    }
+  }
+
+  return 0;
+}
```

Comment:
is this for semver check on integration versions? If so can check: https://stackoverflow.com/questions/6832596/how-can-i-compare-software-version-number-using-javascript-only-numbers. OSD already has the semver package.